rbd.c

```diff
@@ -612,7 +616,177 @@ static int tcmu_rbd_lock_break(struct tcmu_device *dev, char **orig_owner)
 	return ret;
 }
 
-static int tcmu_rbd_lock(struct tcmu_device *dev)
+static int tcmu_rbd_get_lock_tag(struct tcmu_device *dev, uint16_t *tag)
+{
+	struct tcmu_rbd_state *state = tcmu_get_dev_private(dev);
+	char *metadata_owner, *owners[1];
+	size_t num_owners = 1;
+	rbd_lock_mode_t lock_mode;
+	char buf[TCMU_RBD_LCOKER_BUF_LEN];
+	size_t buf_len = TCMU_RBD_LCOKER_BUF_LEN;
+	int ret;
+
+	memset(buf, 0, buf_len);
+
+	ret = rbd_metadata_get(state->image, TCMU_RBD_LOCKER_TAG_KEY,
+			      buf, &buf_len);
+	tcmu_dev_dbg(dev, "get meta got %d %s\n", ret, buf);
+	if (ret)
+		goto done;
+
+	ret = rbd_lock_get_owners(state->image, &lock_mode, owners,
+				  &num_owners);
+	tcmu_dev_dbg(dev, "get lockowner got %d\n", ret);
+	if (ret)
+		goto done;
+	if (!num_owners) {
+		/* there would be stale metadata due to a crash */
+		ret = -ENOENT;
+		goto done;
+	}
+
+	metadata_owner = strstr(buf, "rbd_client=");
+	if (!metadata_owner) {
+		tcmu_dev_err(dev, "Invalid lock tag %s.\n", buf);
+		/* Force initaitor to retry STPG */
+		ret = -ENOENT;
+		goto free_owners;
+	}
+
+	metadata_owner += 11;
+	if (strcmp(metadata_owner, owners[0])) {
+		tcmu_dev_dbg(dev, "owner mismatch %s %s\n", metadata_owner,
+			     owners[0]);
+		/*
+		 * A node could be in the middle of grabbing the lock or it
```

Comment:
Any concern about this potential race? I had originally thought of having each target register a metadata key of something like ```tcmu_target_group_<group id>=<client #>``` just after opening the image. Then, once the lock is acquired, the owner is already in the format of ```client.<client #>``` so you would just need to query the 2 - 4 metadata tags for each group to find the matching client id. Your solution is nice since it only involves a single metadata key lookup.

Reply:
... apparently never mind. I thought that the librbd API provided the client name but it looks like it only provides the IP address of the owner.  I guess that will be something to fix when I add shared lock support (along w/ the ability to store user-provided tags on the lock).